fv3core/utils/grid.py

```diff
@@ -366,14 +367,203 @@ def horizontal_starts_from_shape(self, shape):
         else:
             return 0, 0
 
+    @property
+    def grid_indexing(self) -> "GridIndexing":
+        return GridIndexing.from_legacy_grid(self)
+
+
+class GridIndexing:
+    """
+    Provides indices for cell-centered variables with halos.
+
+    These indices can be used with horizontal interface variables by adding 1
+    to the domain shape along any interface axis.
+    """
+
+    def __init__(
+        self,
+        origin: Index3D,
+        domain: Index3D,
+        n_halo: int,
+        south_edge: bool,
+        north_edge: bool,
+        west_edge: bool,
+        east_edge: bool,
+    ):
+        """
+        Initialize a grid indexing object.
+
+        Args:
+            origin: index of the compute domain origin for cell-centered variables
+            domain: size of the compute domain for cell-centered variables
+            n_halo: number of halo points
+            south_edge: whether the current rank is on the south edge of a tile
+            north_edge: whether the current rank is on the north edge of a tile
+            west_edge: whether the current rank is on the west edge of a tile
+            east_edge: whether the current rank is on the east edge of a tile
+        """
+        self.origin = origin
+        self.domain = domain
+        self.n_halo = n_halo
+        # TODO: split edge booleans into their own class, they aren't needed
+        # for any of the indexing operations, only needed for axis_offsets
+        # and they make the API pretty big
+        # maybe give this responsibility to CubedSphereCommunicator
```

Comment:
I agree there is a TODO here, but I don't think they necessarily need their own class. Once axis_offsets is the only thing using them, we can just pass a CubedSphereCommunicator to axis_offsets. We should preserve the south_edge/west_edge nomenclature as that was a concrete bit of feedback we got from potential users that they would like to see.

Reply:
I think we're on the same page. I'll update the todo to read "into a different class" instead of "into their own class".

>Once axis_offsets is the only thing using them, we can just pass a CubedSphereCommunicator to axis_offsets.

axis_offsets will need a GridIndexing as well as a CubedSphereCommunicator, because the CSC does not have information about the size of the domain (it gets this at call time based on the size of input arguments).

>We should preserve the south_edge/west_edge nomenclature as that was a concrete bit of feedback we got from potential users that they would like to see.

Yes 100%. "giving this responsibility to CubedSphereCommunicator" will mean adding `is_south_edge` etc. methods to CubedSphereCommunicator. We should also refactor the "on_tile_bottom" etc. methods in TilePartitioner.

Reply:
I removed the TODO for now since we added `axis_offsets` to this class. This might make it big enough that we want to split it up later, but I'm not confident enough in that to keep a TODO.